src/Cli/dotnet/LoggerUtility.cs

```diff
@@ -0,0 +1,155 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Logging;
+
+namespace Microsoft.DotNet.Cli
+{
+    internal static class LoggerUtility
+    {
+        public static FacadeLogger? DetermineBinlogger(string[] restoreArgs, string verb)
+        {
+            List<BinaryLogger> binaryLoggers = new();
+
+            for (int i = restoreArgs.Length - 1; i >= 0; i--)
+            {
+                string blArg = restoreArgs[i];
+                if (!IsBinLogArgument(blArg))
+                {
+                    continue;
+                }
+
+                if (blArg.Contains(':'))
+                {
+                    // split and forward args
+                    var split = blArg.Split(':', 2);
+                    var filename = split[1];
+                    if (filename.EndsWith(".binlog"))
+                    {
+                        filename = filename.Substring(0, filename.Length - ".binlog".Length);
+                        filename = $"{filename}-{verb}.binlog";
+                    }
+                    binaryLoggers.Add(new BinaryLogger { Parameters = filename });
+                }
+                else
+                {
+                    // the same name will be used for the build and run-restore-exec steps, so we need to make sure they don't conflict
+                    var filename = $"msbuild-{verb}.binlog";
+                    binaryLoggers.Add(new BinaryLogger { Parameters = filename });
+                }
+
+                // Like in MSBuild, only the last binary logger is used.
+                break;
+            }
+
+            // this binaryLogger needs to be used for both evaluation and execution, so we need to only call it with a single IEventSource across
+            // both of those phases.
+            // We need a custom logger to handle this, because the MSBuild API for evaluation and execution calls logger Initialize and Shutdown methods, so will not allow us to do this.
+            if (binaryLoggers.Count > 0)
+            {
+                var fakeLogger = ConfigureDispatcher(binaryLoggers);
+
+                return fakeLogger;
+            }
+            return null;
+        }
+
+        private static FacadeLogger ConfigureDispatcher(List<BinaryLogger> binaryLoggers)
+        {
+            var dispatcher = new PersistentDispatcher(binaryLoggers);
+            return new FacadeLogger(dispatcher);
+        }
+
+        internal static bool IsBinLogArgument(string arg)
+        {
+            const StringComparison comp = StringComparison.OrdinalIgnoreCase;
+            return arg.StartsWith("/bl:", comp) || arg.Equals("/bl", comp)
+                || arg.StartsWith("--binaryLogger:", comp) || arg.Equals("--binaryLogger", comp)
+                || arg.StartsWith("-bl:", comp) || arg.Equals("-bl", comp);
```

Comment:
eventually we should just create an actual System.CommandLine option for the MSBuild binlog option. We haven't done that yet because it technically supports a [broader grammar than just optional file paths](https://github.com/dotnet/msbuild/blob/6172e6f38e77b53b8ce14754ba637c7ee9896a77/src/Build/Logging/BinaryLogger/BinaryLogger.cs#L429-L444).